tools/stats/metrics.go

```diff
@@ -0,0 +1,241 @@
+	"github.com/filecoin-project/go-lotus/chain/address"
+	"github.com/filecoin-project/go-lotus/chain/types"
+
+	_ "github.com/influxdata/influxdb1-client"
+	models "github.com/influxdata/influxdb1-client/models"
+	client "github.com/influxdata/influxdb1-client/v2"
+)
+
+type PointList struct {
+	points []models.Point
+}
+
+func NewPointList() *PointList {
+	return &PointList{}
+}
+
+func (pl *PointList) AddPoint(p models.Point) {
+	pl.points = append(pl.points, p)
+}
+
+func (pl *PointList) Points() []models.Point {
+	return pl.points
+}
+
+type InfluxWriteQueue struct {
+	influx client.Client
+	ch     chan client.BatchPoints
+}
+
+func NewInfluxWriteQueue(ctx context.Context, influx client.Client) *InfluxWriteQueue {
+	ch := make(chan client.BatchPoints, 128)
+
+	maxRetries := 10
+
+	go func() {
+	main:
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case batch := <-ch:
+				for i := 0; i < maxRetries; i++ {
+					if err := influx.Write(batch); err != nil {
+						log.Printf("Failed to write batch: %w", err)
+						time.Sleep(time.Second * 15)
+						continue
+					}
+
+					continue main
+				}
+
+				log.Printf("Dropping batch due to failure to write")
+			}
+		}
+	}()
+
+	return &InfluxWriteQueue{
+		ch: ch,
+	}
+}
+
+func (i *InfluxWriteQueue) AddBatch(bp client.BatchPoints) {
+	i.ch <- bp
+}
+
+func (i *InfluxWriteQueue) Close() {
+	close(i.ch)
+}
+
+func InfluxClient(addr, user, pass string) (client.Client, error) {
+	return client.NewHTTPClient(client.HTTPConfig{
+		Addr:     addr,
+		Username: user,
+		Password: pass,
+	})
+}
+
+func InfluxNewBatch() (client.BatchPoints, error) {
+	return client.NewBatchPoints(client.BatchPointsConfig{})
+}
+
+func NewPoint(name string, value interface{}) models.Point {
+	pt, _ := models.NewPoint(name, models.Tags{}, map[string]interface{}{"value": value}, time.Now())
+	return pt
+}
+
+func NewPointFrom(p models.Point) *client.Point {
+	return client.NewPointFrom(p)
+}
+
+func RecordTipsetPoints(ctx context.Context, api api.FullNode, pl *PointList, tipset *types.TipSet) error {
+	cids := []string{}
+	for _, cid := range tipset.Cids() {
+		cids = append(cids, cid.String())
+	}
+
+	p := NewPoint("chain.height", int64(tipset.Height()))
+	p.AddTag("tipset", strings.Join(cids, " "))
+	pl.AddPoint(p)
+
+	p = NewPoint("chain.block_count", len(cids))
+	pl.AddPoint(p)
+
+	tsTime := time.Unix(int64(tipset.MinTimestamp()), int64(0))
+	p = NewPoint("chain.blocktime", tsTime.Unix())
+	pl.AddPoint(p)
+
+	for _, blockheader := range tipset.Blocks() {
+		bs, err := blockheader.Serialize()
+		if err != nil {
+			return err
+		}
+
+		p := NewPoint("chain.blockheader_size", len(bs))
+		pl.AddPoint(p)
+	}
+
+	return nil
+}
+
+func RecordTipsetStatePoints(ctx context.Context, api api.FullNode, pl *PointList, tipset *types.TipSet) error {
+	pc, err := api.StatePledgeCollateral(ctx, tipset)
+	if err != nil {
+		return err
+	}
+
+	pcfil := types.BigDiv(pc, types.NewInt(uint64(build.FilecoinPrecision)))
```

Comment:
Change it to:
pcfil := typs.BigDif(pc, types.NewInt(uint64(build.FilecoinPrecision/1000)))
out := pcfi.Float64() / 1000;
```
This way we have some precision on the sub 1FIL level.